### PR TITLE
Sharded Plugin

### LINKED ISSUE #817: Add deepspeed support
Let's support this!

https://github.com/microsoft/DeepSpeed

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #817. Lots of related comments in the issue, but overall fairscale has done a great job of taking the initial DeepSpeed code and making a pytorch module to support the ZERO optimization feature (the main feature from DeepSpeed aside from some custom kernel ops + fp16). Thus I think for a V1, we should offer integration with fairscale and assist in getting DDP changes for model partitioning (model parallel) + await optimisations.

Will require additional tests before merging.

I also note that fairscale install crashes on a remote ubuntu machine. Installing from source however runs fine.

cc @blefaudeux and @ananthsub who mentioned integration already existing internally with lightning, so this PR may unify efforts or be unnecessary. If I could get a lookover this PR I would really appreciate it as well :)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by SeanNaren** [comment]
Going to put some stats here, I've been tracking CUDA memory using a callback inspired by [this](https://github.com/facebookresearch/fairscale/blob/master/benchmarks/benchmark_dataset.py):

```
import torch
from pytorch_lightning import Callback


class CUDACallback(Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        # Reset the memory use counter
        torch.cuda.reset_peak_memory_stats(trainer.global_rank)

        # Dummy training loop
        torch.cuda.synchronize(trainer.global_rank)

    def on_train_epoch_end(self, trainer, pl_module, outputs):
        torch.cuda.synchronize(trainer.global_rank)
        max_memory = torch.cuda.max_memory_allocated(trainer.global_rank) / 2 ** 20

        print(f"[{trainer.global_rank}] : Peak memory {max_memory:.1f}MiB")
```

Running [this transformers script for token classification](https://github.com/huggingface/transformers/blob/master/examples/token-classification/run_pl.sh) I've got some average peak memory stats after the first epoch:

```
Average peak memory allocated after 1 epoch on p3.8xlarge and p3.16xlarge instance types.

4GPUs DDP: 6840.1MiB
8GPUs DDP: 6840.1MiB
4GPUs FairScale OSS: 5263.2MiB (23% memory improvement compared to DDP)
8GPUs FairScale OSS: 4899.03MiB (28.38% memory improvement compared to DDP)
```

**Comment by blefaudeux** [comment]
> Going to put some stats here, I've been tracking CUDA memory using a callback inspired by [this](https://github.com/facebookresearch/fairscale/blob/master/benchmarks/benchmark_dataset.py):
> 
> ```
> import torch
> from pytorch_lightning import Callback
> 
> 
> class CUDACallback(Callback):
>     def on_train_epoch_start(self, trainer, pl_module):
>         # Reset the memory use counter
>         torch.cuda.reset_peak_memory_stats(trainer.global_rank)
> 
>         # Dummy training loop
>         torch.cuda.synchronize(trainer.global_rank)
> 
>     def on_train_epoch_end(self, trainer, pl_module, outputs):
>         torch.cuda.synchronize(trainer.global_rank)
>         max_memory = torch.cuda.max_memory_allocated(trainer.global_rank) / 2 ** 20
> 
>         print(f"[{trainer.global_rank}] : Peak memory {max_memory:.1f}MiB")
> ```
> 
> Running [this transformers script for token classification](https://github.com/huggingface/transformers/blob/master/examples/token-classification/run_pl.sh) I've got some average peak memory stats after the first epoch:
> 
> ```
> Average peak memory allocated after 1 epoch on p3.8xlarge and p3.16xlarge instance types.
> 
> 4GPUs DDP: 6840.1MiB
> 8GPUs DDP: 6840.1MiB
> 4GPUs FairScale OSS: 5263.2MiB (23% memory improvement compared to DDP)
> 8GPUs FairScale OSS: 4899.03MiB (28.38% memory improvement compared to DDP)
> ```

Nice ! FYI there are more savings on the way, it would require wrapping the model and not using DDP (and autograd hook does the job) but that solves the gradient reduce issue and reduces the communications, which is important when multiple nodes are involved. It does not deprecate the version you tested, it just comes on top (if OSS is used with DDP you get what you measured, if OSS is used with a model wrap instead you get a better gradient sharding)

**Comment by SeanNaren** [comment]
Thanks @blefaudeux! Was excited to get numbers for this, really awesome work overall :) 

Are the additional savings only tied to fairscale? Is there a place I can have a look at the code? Would be good to get a head start on figuring out API changes to support this

**Comment by blefaudeux** [comment]
> Thanks @blefaudeux! Was excited to get numbers for this, really awesome work overall :)
> 
> Are the additional savings only tied to fairscale? Is there a place I can have a look at the code? Would be good to get a head start on figuring out API changes to support this

No PR yet, the work is in https://github.com/facebookresearch/fairscale/tree/oss_autograd
The interface could be 
```
model = myAwesomeModel()
optimizer = OSS(*optimizer_params)
model = ShardedDDP(model, optimizer, *some_basic_params)
```
and then a normal training loop with optimizer and model. Gradients are automatically reduced to the right rank and the optimizer state and gradients are sharded, which shaves some more memory from what you have right now.

What do you think ?

**Comment by SeanNaren** [comment]
Yep interface makes sense to me. I also see the model dispatch making its way into this branch which is super cool!

I don't think it makes sense to support both DDP + OSS and Sharded DDP + OSS since we'll need to install fairscale regardless and Sharded DDP seems like the successor. I think it makes sense to go forward ShardedDDP and have it replace the current implementation, thoughts here?

**Comment by blefaudeux** [comment]
> Yep interface makes sense to me. I also see the model dispatch making its way into this branch which is super cool!
> 
> I don't think it makes sense to support both DDP + OSS and Sharded DDP + OSS since we'll need to install fairscale regardless and Sharded DDP seems like the successor. I think it makes sense to go forward ShardedDDP and have it replace the current implementation, thoughts here?

It may require some benchmarking, basically one feature from DDP which is hard to replicate is the overlap in between BW and reduce, the gradients in DDP are all-reduced step by step when walking back the graph. Currently what's "easy" to do is FW/BW/reduce (not all reduce), but the overlap is lost. When we shard the model (next steps, a little more involving) it's not that much of an issue because we can overlap the reduce of the lower shard with the BW of the upper one.

From what I can see, currently (state+gradient sharding):
- on a single node OSS + DDP is a little faster than OSS+ad-hoc-reduce but consumes more memory
- when multiple nodes are involved, then OSS+ad-hoc-reduce wins it all (faster and less memory)

**Comment by SeanNaren** [comment]
Yeah still not sure it's worth supporting the DDP version. Having the DDP lightning wrapper just for a small benefit in speed for single node setups (and personally I think what's more important is reducing memory allocation) where the cap of total GPUs is small, I don't think is necessary and adds confusion unless I'm mistaken!

**Comment by SeanNaren** [comment]
Pushing WIP changes integrating ShardedDDP using the `oss_autograd` fairscale branch,  in a similar fashion to vanilla DDP. Unfortunately I had to do quite a bit of overriding due since the model requires multiple arguments (not just the input to the model, but multi-input/targets for forward/loss calculation).

Running into an issue using multiple GPUs where training hangs, currently investigating this. Also seeing if there is a nicer way to handle requiring grads for the input. Main issue being I've made the assumption (for now) that the inputs are tuples within the batch i.e within ModelDispatch:

```python
# All inputs need to required_grad for autograd to properly track the first dispatch layer
# Will currently break if a dict or something, may require a recursive check
if isinstance(inputs, tuple):
     for i in inputs:
           i.requires_grad = True
```

**Comment by SeanNaren** [comment]
@williamFalcon I've been trying to keep up with the plugin API which is neat! I still think the fairscale integration should live as a native accelerator, because there are more features to come from fairscale, but was curious on your thoughts.

**Comment by SeanNaren** [comment]
A few updates to track here, running into issues supporting kwarg input for forward/backward pass. Seems a solution was discussed here: https://github.com/facebookresearch/fairscale/issues/160#issuecomment-717580898 and here: https://github.com/facebookresearch/fairscale/issues/160#issuecomment-718035415

Seems like there are some performance issues when moving to multi-node which I haven't been able to test that ben has reported: https://github.com/facebookresearch/fairscale/pull/157#issuecomment-718132892

Since I don't think its worth longer term to just get the OSS optimizer into lightning without SDP, we'll put this on hold till performance issues and function args are solved!

**Comment by ananthsub** [comment]
> @williamFalcon I've been trying to keep up with the plugin API which is neat! I still think the fairscale integration should live as a native accelerator, because there are more features to come from fairscale, but was curious on your thoughts.

naming nit: can we name this as sharded ddp accelerator? fairscale as a library will eventually have more components we may want to plug in elsewhere

**Comment by SeanNaren** [comment]
@ananthsub I was thinking about that... even ZeRO is better or Deepspeed, but I don't mind just calling it ShardedDDP

**Comment by SeanNaren** [comment]
Just an update, lots of kinks ironed out with Ben's upstream changes with now easy support for whatever args need to go through the lightning module steps. Here's some further results:

_Benchmarking Text Classification pipeline using the [HF/Lightning Transformers script](https://github.com/SeanNaren/transformers/blob/feature/wip2/examples/text-classification/run_pl.sh)._ 

_Parameters are slightly different: BERT large cased + FP16, Batch Size 32 using a p3.16xlarge AWS node_

| Backend                     | GPUs | Average Peak Mem (MiB) | Epoch Time(s) |
|-----------------------------|------|------------------------|---------------|
| Single GPU                  |    1 |                12738.6 |         38.98 |
| DDP                         |    4 |                14011.9 |         11.80 |
| DDP                         |    8 |                14011.9 |          6.13 |
| FairScale OSS + Sharded DDP |    4 |                 9892.9 |         11.85 |
| FairScale OSS + Sharded DDP |    8 |                 9420.9 |          5.98 |

Currently only measured for one node atm, need to look at multi-node but a 39% memory improvement for marginally better speed with a flag change (`accelerator=ddp` to `accelerator=ddp_sharded`) is absolutely ridiculous, awesome work @blefaudeux :)

**Comment by blefaudeux** [comment]
> @ananthsub I was thinking about that... even ZeRO is better or Deepspeed, but I don't mind just calling it ShardedDDP

FYI the naming within PyTorch is kind of a moving target, but it will probably not be ZeRO, deemed a "brand" which belongs to MSFT.

**Comment by SeanNaren** [comment]
We're nearly there... awaiting on a bunch of PRs to be merged into fairscale and then hopefully a release after.

From the PL side, I need to integrate custom clip grad logic from fairscale, and in a separate issue consider the interface to provide plugins via strings to the trainer.

File: pytorch_lightning/overrides/fairscale.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from fairscale.nn.data_parallel.sharded_ddp import ShardedDataParallel
+from fairscale.optim import OSS
+from torch import nn
+
+from pytorch_lightning.utilities import rank_zero_warn
+
+
+class LightningOSS(OSS):
+
+    def state_dict(self) -> Dict[str, Any]:
```
Comment (ananthsub) [inline pytorch_lightning/overrides/fairscale.py:25]: n00b question: why do we override the state dict? how do we ensure this stays in sync with upstream fairscale?

File: pytorch_lightning/overrides/fairscale.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        return {
+            "state": [s["state"] for s in self._all_states],
+            "param_groups": param_groups,
+            "partition": partition,
+            "local_state_dict": False,
+        }
+
+
+class LightningShardedDataParallel(ShardedDataParallel):
+    def __init__(
```
Comment (ananthsub) [inline pytorch_lightning/overrides/fairscale.py:63]: same with this, I think this approach will run into the DDP issues we see now mentioned here: https://github.com/PyTorchLightning/pytorch-lightning/issues/4630

File: pytorch_lightning/overrides/fairscale.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            self,
+            base_model: nn.Module,
+            sharded_optimizer: Union[OSS, List[OSS]],
```
Comment (ananthsub) [inline pytorch_lightning/overrides/fairscale.py:66]: should the OSS wrapping be automatic? or should users explicitly use OSS in `configure_optimizers` of their lightning module?

Comment (SeanNaren) [reply pytorch_lightning/overrides/fairscale.py:66]: I think atm it supports both. We just skip wrapping the optimizer in the sharded_plugin if the user has already wrapped it!

File: pytorch_lightning/overrides/fairscale.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/overrides/fairscale.py:22]: I spoke to ben and it doesn't make sense to allow override of the forward logic as this can lead to some side effects with reduce handles etc. The override is extremely minimal now, so I think for an initial release, this is fine!
